web/api/v1/api.go

```diff
@@ -805,6 +805,58 @@ func rulesAlertsToAPIAlerts(rulesAlerts []*rules.Alert) []*Alert {
 	return apiAlerts
 }
 
+type metadata struct {
+	Type textparse.MetricType `json:"type"`
+	Help string               `json:"help"`
+	Unit string               `json:"unit"`
+}
+
+func (api *API) metricMetadata(r *http.Request) apiFuncResult {
+	metrics := map[string]map[metadata]struct{}{}
+
+	limit := -1
+	if s := r.FormValue("limit"); s != "" {
+		var err error
+		if limit, err = strconv.Atoi(s); err != nil {
+			return apiFuncResult{nil, &apiError{errorBadData, errors.New("limit must be a number")}, nil, nil}
+		}
+	}
+
+	for _, tt := range api.targetRetriever.TargetsActive() {
+		for _, t := range tt {
+			for _, mm := range t.MetadataList() {
+				m := metadata{Type: mm.Type, Help: mm.Help, Unit: mm.Unit}
+				ms, ok := metrics[mm.Metric]
+
+				if !ok {
+					ms = map[metadata]struct{}{}
+					metrics[mm.Metric] = ms
+				}
+
+				ms[m] = struct{}{}
+			}
+		}
+	}
+
+	res := map[string][]metadata{}
```

Comment:
I'd add a comment that this is needed for marshaling.

Reply:
I'll add it in the next one which should be up soon.